### PR TITLE
ddl: support drop sequence in TiDB

### PR DESCRIPTION
**Pull request by AilinKid**
support drop sequence in TiDB.
```
DROP [TEMPORARY] SEQUENCE [IF EXISTS] sequence_name [, sequence_name] ...
```

### What is changed and how it works?
1. drop sequence share many logic code with `dropTableOrView`, so I merge them together.
2. drop sequence will also delete the sequence value kv pair (drop table and drop table will also do this. But don't worry, `HDel()` won't take effect if the sequence kv-pair is not exist in case of drop table/view)

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has interface methods change

Related changes

 - Need to update the documentation

Release note

 - ddl: Support drop sequence in TiDB.

### CODE REVIEW

File: ddl/sequence_test.go [hunk 1]
```diff
@@ -84,3 +84,37 @@ func (s *testSequenceSuite) TestCreateSequence(c *C) {
 	c.Assert(err, NotNil)
 	c.Assert(err.Error(), Equals, "[planner:1142]CREATE command denied to user 'localhost'@'myuser' for table 'my_seq'")
 }
+
+func (s *testSequenceSuite) TestDropSequence(c *C) {
+	s.tk = testkit.NewTestKit(c, s.store)
```
[the remark this answers is not in the document]
Comment (AilinKid) [reply ddl/sequence_test.go]: good catch.

**Comment by AilinKid** [comment]
> It does not look like privilege checks are being done here. I assume it will be handled in a followup PR?

Good catch! thanks for reminding me.

File: executor/ddl.go [hunk 2]
```diff
@@ -251,9 +253,26 @@ func isSystemTable(schema, table string) bool {
...
 func (e *DDLExec) executeDropTableOrView(s *ast.DropTableStmt) error {
+	obt := tableObject
+	if s.IsView {
+		obt = viewObject
+	}
+	return e.dropTableViewSequence(s.Tables, obt, s.IfExists)
+
+}
+
+func (e *DDLExec) dropTableViewSequence(objects []*ast.TableName, obt objectType, ifExists bool) error {
```
Comment (bb7133) [inline executor/ddl.go]: The naming is weird. How about renaming it to `dropTableObject` and add some comments?

File: executor/ddl.go [hunk 3]
```diff
@@ -251,9 +253,26 @@ func isSystemTable(schema, table string) bool {
...
 
+type objectType int
+
+const (
+	tableObject objectType = iota
+	viewObject
+	sequenceObject
+)
+
 func (e *DDLExec) executeDropTableOrView(s *ast.DropTableStmt) error {
```
Comment (bb7133) [inline executor/ddl.go]: Please separate this function to `executeDropTable` and `executeDropView`(just what you did for Sequences).

Both of them can simply call `dropTableViewSequence`

File: executor/ddl.go (hunk 2 above)
Comment (AilinKid) [reply executor/ddl.go]: nice, addressed.

File: executor/ddl.go (hunk 3 above)
Comment (AilinKid) [reply executor/ddl.go]: Got it, addressed.

**Comment by AilinKid** [comment]
> It does not look like privilege checks are being done here. I assume it will be handled in a followup PR?

I've added the check, please take a look~

File: executor/ddl.go [hunk 4]
```diff
@@ -251,9 +257,29 @@ func isSystemTable(schema, table string) bool {
...
+
+func (e *DDLExec) executeDropView(s *ast.DropTableStmt) error {
+	return e.dropTableObject(s.Tables, viewObject, s.IfExists)
+}
+
+func (e *DDLExec) executeDropSequence(s *ast.DropSequenceStmt) error {
+	return e.dropTableObject(s.Sequences, sequenceObject, s.IfExists)
+}
+
+func (e *DDLExec) dropTableObject(objects []*ast.TableName, obt objectType, ifExists bool) error {
```
Comment (bb7133) [inline executor/ddl.go]: Please add some comments here...the function actually applies to `tableObject`, `viewObject` and `sequenceObject`.

File: ddl/sequence_test.go [hunk 5]
```diff
@@ -84,3 +84,70 @@ func (s *testSequenceSuite) TestCreateSequence(c *C) {
...
+	s.tk.MustExec("create sequence seq")
+	_, err := s.tk.Exec("drop sequence seq, seq2")
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Equals, "[schema:4139]Unknown SEQUENCE: 'test.seq2'")
+
+	// Test the specified object is not sequence.
+	s.tk.MustExec("create table seq3 (a int)")
+	_, err = s.tk.Exec("drop sequence seq3")
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Equals, "[ddl:1347]'test.seq3' is not SEQUENCE")
```
Comment (tangenta) [inline ddl/sequence_test.go]:
```suggestion
	c.Assert(terror.ErrorEqual(err, ddl.ErrWrongObject), IsTrue)
```
Thus we can change the error message(in the future) without updating this test. It is also safe to pass in a `nil` error.

File: ddl/sequence_test.go [hunk 6]
```diff
@@ -84,3 +84,70 @@ func (s *testSequenceSuite) TestCreateSequence(c *C) {
...
+	c.Assert(err, IsNil)
+	_, err = s.tk.Exec("drop sequence seq")
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Equals, "[schema:4139]Unknown SEQUENCE: 'test.seq'")
+
+	// Test drop table when the object is a sequence.
+	s.tk.MustExec("create sequence seq")
+	_, err = s.tk.Exec("drop table seq")
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Equals, "[schema:1051]Unknown table 'test.seq'")
```
Comment (tangenta) [inline ddl/sequence_test.go]: Is it expected? MariaDB has different behavior.
```
mariadb root@172.17.0.3:test> create sequence s;
Query OK, 0 rows affected
Time: 0.019s
mariadb root@172.17.0.3:test> drop table s
Query OK, 0 rows affected
Time: 0.013s
```

Comment (AilinKid) [reply ddl/sequence_test.go]: I've talked with wangcong about this diff. we think drop table shouldn't influence sequence object.

File: ddl/sequence_test.go (hunk 5 above)
Comment (AilinKid) [reply ddl/sequence_test.go]: Good, make sense~

**Comment by AilinKid** [comment]
> In MariaDB,
> 
> ```
> mariadb root@172.17.0.3:test> show create table t
> 1 row in set
> Time: 0.006s
> +-------+----------------------------------------+
> | Table | Create Table                           |
> +-------+----------------------------------------+
> | t     | CREATE TABLE `t` (                     |
> |       |   `a` int(11) DEFAULT NULL             |
> |       | ) ENGINE=InnoDB DEFAULT CHARSET=latin1 |
> +-------+----------------------------------------+
> mariadb root@172.17.0.3:test> drop sequence if exists t
> Query OK, 0 rows affected
> Time: 0.001s
> ```
> 
> Perhaps skip `ErrWrongObject` if `if exists` is specified.

Exactly right, I've fixed it, PTAL

File: meta/meta.go [hunk 7]
```diff
@@ -400,6 +400,9 @@ func (m *Meta) DropTableOrView(dbID int64, tblID int64, delAutoID bool) error {
 			return errors.Trace(err)
 		}
 	}
+	if err := m.txn.HDel(dbKey, m.sequenceKey(tblID)); err != nil {
```
Comment (zimulala) [inline meta/meta.go]: Could we reduce this operation when doing `drop table/view`? We can add a new function.

File: ddl/ddl_worker.go [hunk 8]
```diff
@@ -531,7 +531,7 @@ func (w *worker) runDDLJob(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64,
 		ver, err = onRepairTable(d, t, job)
 	case model.ActionCreateView:
 		ver, err = onCreateView(d, t, job)
-	case model.ActionDropTable, model.ActionDropView:
+	case model.ActionDropTable, model.ActionDropView, model.ActionDropSequence:
 		ver, err = onDropTableOrView(t, job)
```
Comment (zimulala) [inline ddl/ddl_worker.go]: Do we need to update these functions like `onDropTableOrView`?

File: meta/meta.go (hunk 7 above)
Comment (AilinKid) [reply meta/meta.go]: Addressed.

**Pull request merged**